p4-16/psa/PSA.mdk

```diff
@@ -245,76 +245,185 @@ Operations in the egress pipeline will not execute if drop is called in ingress.
+struct psa_ingress_metadata_for_parser_t {
+   bool is_recirc;
+}
 ```
 
-#### Clone with extra data operation { #sec-pre-clone-with-data }
-
-Create a copy of the packet with additional data and send it to the
-specified port.
-
-The PSA specifies four types of cloning, with the packet sourced
-from different points in the pipeline and sent back to ingress
-or to the buffering queue in the egress (see [#sec-cloning-methods]).
-\TODO: needs both source and destination data
-
-Parameters:
-
-- clone_method  The type of cloning.
-- port          The port to send the cloned packet to.
-- data additional header data attached to the packet
-
 ```
-void clone<T>  (in CloneMethod_t clone_method, in PortId_t port, in T data);
+struct psa_ingress_metadata_for_pre_t {
+   bool clone_ena;
+   CloneSpectT clone_spec;
+   bool resubmit_ena;
+}
 ```
 
-#### Resubmit operation
-
-Send a packet to the ingress port with additional data appended.
-
-This operation is intended for recursive packet processing.
-\TODO: needs both source and destination data
-
-Parameters:
-
-- data A header definition that can be added to the set of
-  packet headers.
-- port The input port at which the packet will be resubmitted.
-
 ```
-void resubmit<T>(in T data, in PortId_t port);
+struct psa_egress_metadata_for_pre_t {
+   bool clone_ena;
+   CloneSpecT clone_spec;
+   bool recirc_ena;
+}
 ```
 
-#### Recirculate operation
-
-Send a post deparse packet to the ingress port with additional data
-appended.
-
-This operation is intended for recursive packet processing.
-\TODO: needs both source and destination data
-
-Parameters:
-
-- data A header definition that can be added to the set of packet headers.
-- port The input port at which the packet will be resubmitted.
+#### Clone
+
+A PSA implementation provides a clone mechanism to create a copy of
+the original packet as an independent packet instance. The cloned
+packet can be a duplicate of the original packet or a copy of the
+deparsed packet after egress pipeline. The clone mechanism can submit
+the cloned packet to ingress parser or buffering mechanism. In a PSA
+implementation, the clone mechanism is implemented in PRE.
+
+The clone mechanism can optionally attach metadata to the cloned
+packet. A PSA implementation provides the `clone` extern to specify
+the attached metadata. The `clone` extern provides a `emit` method
+which accepts the attached metadata of generic type `T`. In a PSA
+implementation, the metadata is prepended to the cloned packet. It is
+the responsibility of the programmer to parse the cloned packet
+correctly to extract the attached metadata. The attached metadata can
+be of type `header`, `header stack`, `header union` or `struct` of the
+above types. Invoking the `emit` method multiple times will attach all
+specified metadata to the same cloned packet. The PSA architecture
+instantiates the `clone` extern in the ingress and egress deparser. A
```

Comment:
The good news is that clone extern does not have a constructor, so it will be impossible to instantiate even by mistake.

Strictly speaking, the architecture doesn't instantiate it **in** the deparsers, but outside of them and then simply passes the extern to the deparsers, similar to how it does with packet_out.